### DISCUSSION TITLE
add global await

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#6520] [Language Ideas]
### Background and motivation

I suggest adding block 
```
await{
}
```
For asynchronous methods in this block, await will be supplied exactly.
I have given two methods below, they should work the same

### API Proposal

```csharp
  public async Task ExampleNotAwait()
        {
            await{
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
            }
        }
```

### API Usage

```csharp
    class Example
    {
        public async Task ExampleWithAwait()
        {
            await Task.Delay(100);
            await Task.Delay(100);
            await Task.Delay(100);
            await Task.Delay(100);
            await Task.Delay(100);
        }
        public async Task ExampleNotAwait()
        {
            await{
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
                Task.Delay(100);
            }
        }
    }
```

**Comment by Mr0N** [comment]
```
public async Task<string> Get()
        {
            await{
                var client = new HttpClient();
               return client.GetAsync("")
                    .Content.
                    ReadAsStringAsync();
            }
        }
```

**Comment by dnperfors** [comment]
I think this suggestion is coming from an F# developer, where this is rather normal approach of doing asynchronous work. However, I don't think just providing this syntactic sugar is worth the hassle.

If you would suggest introducing monads, that might be a better proposal and then this syntax could be part of it. (Although I doubt that it will ever make it into C#)

**Reply by Mr0N** [reply]
> I think this suggestion is coming from an F# developer, where this is rather normal approach of doing asynchronous work. However, I don't think just providing this syntactic sugar is worth the hassle.
> 
> If you would suggest introducing monads, that might be a better proposal and then this syntax could be part of it. (Although I doubt that it will ever make it into C#)

No, I'm not an F# developer, I don't really like F# at all)
Just why put await around each asynchronous method if the compiler can do it.

**Reply by En3Tho** [reply]
In F# awaiting is an explicit thing too - you have to use "let!/do!" keywords to indicate your async/task expression that this one should be awaited.

```fsharp
task {
    do! Task.Delay(1000)
    let! x = Task.FromResult(1)
    ...
}
```

 I don't think it's possible to create F# builder that supports auto-awaiting without any special hints

**Reply by Mr0N** [reply]
> 

Maybe there is a second way, I suggest not to write await manually, but to substitute them by the compiler

**Reply by Mr0N** [reply]
The compiler seems to know where the asynchronous method is, why can't it substitute await?

**Reply by huoyaoyuan** [reply]
> The compiler seems to know where the asynchronous method is

Awaitable methods are much more complex than you expected.
Any type with proper `GetAwaiter()` can be awaited. The `GetAwaiter()` method can be extension method, and this is how `ValueTask`, `ConfigureAwait()` and WinRT awaitable works.
If the compiler adds `await` for every awaitable type, adding such extension method on a type can break any code massively.

**Reply by Mr0N** [reply]
> > The compiler seems to know where the asynchronous method is
> 
> Awaitable methods are much more complex than you expected. Any type with proper `GetAwaiter()` can be awaited. The `GetAwaiter()` method can be extension method, and this is how `ValueTask`, `ConfigureAwait()` and WinRT awaitable works. If the compiler adds `await` for every awaitable type, adding such extension method on a type can break any code massively.

![image](https://user-images.githubusercontent.com/61237140/194884940-b6d76b07-2ffa-4d6e-a634-ffb4a1324933.png)

**Reply by Mr0N** [reply]
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.asyncstatemachineattribute?view=net-6.0

**Reply by HaloFour** [reply]
The compiler definitely could do this automatically, but it was a very intentional design decision by the language team that it not do this.  They want the asynchronous boundaries to be obvious when reading the code.  IMO wrapping any given body of code in a block that automatically applies `await` to any "awaitable" expression would muddy that intent.

**Reply by HaloFour** [reply]
`AsyncStateMachine` is not how the compiler recognizes an expression as "awaitable":

https://devblogs.microsoft.com/pfxteam/await-anything/

**Reply by Mr0N** [reply]
> The compiler definitely could do this automatically, but it was a very intentional design decision by the language team that it not do this. They want the asynchronous boundaries to be obvious when reading the code. IMO wrapping any given body of code in a block that automatically applies `await` to any "awaitable" expression would muddy that intent.

Well, it was in 2012, when there was not so much async code, in 2022 there is a lot of it, and writing on each await method is not very pleasant and, in my opinion, superfluous

**Reply by CyrusNajmabadi** [reply]
In 2022 I still like async boundaries being explicit.

**Reply by HaloFour** [reply]
> Well, it was in 2012, when there was not so much async code, in 2022 there is a lot of it, and writing on each await method is not very pleasant and, in my opinion, superfluous

Time and popularity haven't changed anything regarding the design decisions here.  It was expected (and hoped) that `async`/`await` would be popular and widely used.  In any conversation about this subject the team has reaffirmed that they want these asynchronous boundaries to be intentional and obvious in the code.

**Reply by Mr0N** [reply]
> > Well, it was in 2012, when there was not so much async code, in 2022 there is a lot of it, and writing on each await method is not very pleasant and, in my opinion, superfluous
> 
> Time and popularity haven't changed anything regarding the design decisions here. It was expected (and hoped) that `async`/`await` would be popular and widely used. In any conversation about this subject the team has reaffirmed that they want these asynchronous boundaries to be intentional and obvious in the code.

Well, if you add the suffix Async to the method, then it is clear where the asynchronous methods are, you can also make a second highlight, hide everything, there will be boundaries and you will not need to write await around each method

**Reply by HaloFour** [reply]
The fact that the method might be `async` (and that convention doesn't necessarily describe _how_ it's async), that does not indicate where the asynchronous boundaries are.  And you can `await` much more than method calls.

**Reply by Mr0N** [reply]
> The fact that the method might be `async` (and that convention doesn't necessarily describe _how_ it's async), that does not indicate where the asynchronous boundaries are. And you can `await` much more than method calls.

well, I'm not saying to remove await as it is now, but simply to add a new function

**Reply by dnperfors** [reply]
> writing on each await method is not very pleasant and, in my opinion, superfluous

I'd rather write a couple of await keywords than to have my code extra indented.
Personally I don't see the problem here. If I look at my current project, I hardly specify await myself. Of course there some places where you need them, but most of the time I just pass the complete task.

**Reply by Xyncgas** [reply]
> In F# awaiting is an explicit thing too - you have to use "let!/do!" keywords to indicate your async/task expression that this one should be awaited.
> 
> ```fsharp
> task {
>     do! Task.Delay(1000)
>     let! x = Task.FromResult(1)
>     ...
> }
> ```
> 
> I don't think it's possible to create F# builder that supports auto-awaiting without any special hints

That's because it's happening inside an async CE not a CE called await

**Reply by mikernet** [reply]
@dnperfors 
> most of the time I just pass the complete task.

Just FYI, that's not actually equivalent to awaiting the task in your method, and as a general best practice should be avoided in most cases. One reason I can think of is that it makes debugging much harder because stack traces in the task will no longer contain the caller. The one exception I can think of is when you have method overloads in the same class it's ok to return the task directly because the stack trace will at least still mostly make sense.

Example:

```cs
public async Task MethodAsync()
{
    throw new InvalidOperationException();
}

public Task NoAwaitAsync()
{
    return MethodAsync();
}

public async Task AwaitAsync()
{
    await MethodAsync();
}
```
If you call `NoAwaitAsync`, your exception stack trace looks like:
```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Program.Test.MethodAsync() in :line 5
   at Program.<<Initialize>>d__0.MoveNext() in :line 24
```
Notice there is nothing there indicating that `NoAwaitAsync` was any way involved so you have no idea what called `MethodAsync` to cause the problem. If you call `AwaitAsync` instead, you get this much nicer stack trace:
```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Program.Test.MethodAsync() in :line 5
   at Program.Test.AwaitAsync() in :line 15
   at Program.<<Initialize>>d__0.MoveNext() in :line 33
```

**Comment by Xyncgas** [comment]
We definitely see things being more simple inside F#'s async CE which's mentioned by some one,
if we declare a scope of await then everything inside will be awaited, this is both explicit and very much a syntax sugar,
it's the equivalent of telling your IDE to paste the word await on the beginning of these lines which you can already do so even if you are worried 'the end of the world' that there's some danger in enabling this the damage is already done

**Reply by Mr0N** [reply]
Well, firstly, the compiler will substitute the keyword only in the await{} block.
What can be the harm? Asynchronous code will look like synchronous, hide everything

**Reply by HaloFour** [reply]
> What can be the harm? Asynchronous code will look like synchronous, hide everything

That is the harm, the team very much does not want asynchronous code to look like synchronous code, or to hide anything.

**Reply by CyrusNajmabadi** [reply]
While i cannot speak for everyone, i can say that i haven't noticed any changing here among the team about this aspect of async programming.  The explicitness remains a virtue that we don't seem to be changing our opinions on.

**Reply by Mr0N** [reply]
> > What can be the harm? Asynchronous code will look like synchronous, hide everything
> 
> That is the harm, the team very much does not want asynchronous code to look like synchronous code, or to hide anything.

Well, it will not look like synchronous because it will be in the await block.
Why doesn't the team want asynchronous code to look like synchronous?

**Reply by HaloFour** [reply]
> Why doesn't the team want asynchronous code to look like synchronous?

Because it doesn't behave like synchronous code.  The fact that the execution may resume on a separate thread is reason enough, but it also impacts what language structures you can or can't use around the boundary.  The explicitness of the boundary is a virtue, and it's very unlikely that throwing it within some kind of scoped block would be considered enough to reverse that design decision.

**Reply by theunrepentantgeek** [reply]
Seeing the **await** keyword is a strong clue that there is serious complexity underlying that call and a reminder to any dev reading the code that they need to be alert. For one, failure modes for asynchronous code are very different to those for synchronous code.

**Reply by CyrusNajmabadi** [reply]
> Why doesn't the team want asynchronous code to look like synchronous?

Because it's not synchronous, and synchronous vs asynchronous have very different semantics that we like clearly delineated.  For myself, it's critical to see where those yield points are as it literally can and will be the case that other code can (and will) run in that same sync-context, and thus may affect the state of the program when the yield point resumes.  It requires deep thought on the impact of that, and should not happen implicitly for me.

**Comment by Mr0N** [comment]
```
public async Task<string> Get()
        {

                var client = new HttpClient();
               return async await client.GetAsync("")
                    .Content.
                    ReadAsStringAsync();
        }
```
Well, then add something like this, if async await is placed at the beginning of the expression, then await will be substituted for this expression in all expressions
```
public async Task<string> Get()
        {

                var client = new HttpClient();
               return await((await client.GetAsync(""))
                    .Content.
                    ReadAsStringAsync());
        }
```
analogue of the code above

**Reply by HaloFour** [reply]
This looks so much nicer than both of those:

```cs
public async Task<string> Get()
{
    var client = new HttpClient();
    var response = await client.GetAsync("");
    return await response.Content.ReadAsStringAsync();
}
```

**Reply by Mr0N** [reply]
Here are two lines of code, there is one. It’s not so easy to add that to the expressed 10 asynchronous methods, and all of them must be called
`
var result = await (await (await (await obj.MethodAsync).MethodAsync).MethodAsync);
var result2 = async await obj.MethodAsync().MethodAsync().MethodAsync();
`

**Reply by CyrusNajmabadi** [reply]
i would far prefer:

```c#
var v1 = await obj.MethodAsync();
var v2 = await v1.MethodAsync();
var v3 = await v2.MethodAsync();
```

It's clear and readable what's going on, and all the await points are perfectly understandable.

With:

```c#
var result2 = async await obj.MethodAsync().MethodAsync().MethodAsync();
```

I have no clue at what point yields are actually happening.

**Reply by CyrusNajmabadi** [reply]
> ```c#
> var obj1 = await obj.MethodAsync();
> var obj2 = await obj1.MethodAsync();
> var obj3 = await obj2.MethodAsync();
> var obj4 = await obj3.MethodAsync();
> ```
> 
> Is it so good?

Yes, aside from naming things well, this is super clear code that is easy to write and maintain.

**Reply by Mr0N** [reply]
> i would far prefer:
> 
> ```cs
> var v1 = await obj.MethodAsync();
> var v2 = await v1.MethodAsync();
> var v3 = await v2.MethodAsync();
> ```

Can Linq be written in the same way?

**Reply by HaloFour** [reply]
> Here are two lines of code, there is one.

It's not a goal of the language to reduce the number of lines of code as much as possible.

And I think the following is still much clearer:

```cs
var result1 = await obj.MethodAsync;
var result2 = await result1.MethodAsync;
var result3 = await result2.MethodAsync;
var result4 = await result3.MethodAsync;
```

The team has expressed a potential interest in exploring a general way of piping or chaining expressions which could include a way to `await` within an expression, but the existing proposals don't seem to have gathered much steam.  They might be willing to look more into something like that than anything that makes `await` implicit, even scoped to a single expression.

**Reply by CyrusNajmabadi** [reply]
> Can Linq be written in the same way?

I don't know what this means.  Can you clarify?
